File: crates/breez-sdk/bindings/examples/cli/langs/python/src/breez_cli/webhooks.py
import argparse

from breez_sdk_spark import (
    RegisterWebhookRequest,
    UnregisterWebhookRequest,
    WebhookEventType,
)

from breez_cli.serialization import print_value

# Webhooks subcommand names (used for REPL completion)
WEBHOOKS_COMMAND_NAMES = [
    "webhooks register",
    "webhooks unregister",
    "webhooks list",
]


def _parser(name, description=""):
    return argparse.ArgumentParser(prog=f"webhooks {name}", description=description)


def _parse_event_type(s):
    mapping = {
        "lightning-receive": WebhookEventType.LIGHTNING_RECEIVE_FINISHED,
        "lightning-send": WebhookEventType.LIGHTNING_SEND_FINISHED,
        "coop-exit": WebhookEventType.COOP_EXIT_FINISHED,
        "static-deposit": WebhookEventType.STATIC_DEPOSIT_FINISHED,
    }
    result = mapping.get(s.lower())
    if result is None:
        raise ValueError(
            f"Unknown event type: {s}. Valid values: lightning-receive, "
            "lightning-send, coop-exit, static-deposit"
        )
    return result


# --- register ---

def _build_register_parser():
    p = _parser("register", "Register a new webhook")
    p.add_argument("url", help="URL that will receive webhook notifications")
    p.add_argument("secret", help="Secret for HMAC-SHA256 signature verification")
    p.add_argument("events", nargs="+",
                   help="Event types (lightning-receive, lightning-send, coop-exit, static-deposit)")
    return p

async def _handle_register(sdk, args):
    event_types = [_parse_event_type(e) for e in args.events]
    result = await sdk.register_webhook(
        request=RegisterWebhookRequest(
            url=args.url,
            secret=args.secret,
            event_types=event_types,
        )
    )
    print_value(result)


# --- unregister ---

def _build_unregister_parser():
    p = _parser("unregister", "Unregister a webhook")
    p.add_argument("webhook_id", help="ID of the webhook to unregister")
    return p

async def _handle_unregister(sdk, args):
    await sdk.unregister_webhook(
        request=UnregisterWebhookRequest(webhook_id=args.webhook_id)
    )
    print("Webhook unregistered successfully")


# --- list ---

def _build_list_parser():
    return _parser("list", "List all registered webhooks")

async def _handle_list(sdk, _args):
    result = await sdk.list_webhooks()
    print_value(result)


# ---------------------------------------------------------------------------
# Registry and dispatch
# ---------------------------------------------------------------------------

def _build_webhooks_registry():
    return {
        "register": (_build_register_parser(), _handle_register),
        "unregister": (_build_unregister_parser(), _handle_unregister),
        "list": (_build_list_parser(), _handle_list),
    }


_REGISTRY = None

def _get_registry():
    global _REGISTRY
    if _REGISTRY is None:
        _REGISTRY = _build_webhooks_registry()
    return _REGISTRY


async def dispatch_webhooks_command(args, sdk):
    """Dispatch a webhooks subcommand given the args after 'webhooks'."""
    registry = _get_registry()

    if not args or args[0] == "help":
        print("\nWebhooks subcommands:\n")
        for name, (parser, _) in sorted(registry.items()):
            desc = parser.description or ""
            print(f"  webhooks {name:30s} {desc}")
        print()
        return

    sub_name = args[0]
    sub_args = args[1:]

    if sub_name not in registry:
        print(f"Unknown webhooks subcommand: {sub_name}. Use 'webhooks help' for available commands.")
        return

    parser, handler = registry[sub_name]
    try:
        parsed = parser.parse_args(sub_args)
    except SystemExit:
        return

    await handler(sdk, parsed)

File: crates/breez-sdk/bindings/examples/cli/langs/flutter/lib/webhooks.dart
import 'package:breez_sdk_spark_flutter/breez_sdk_spark.dart';

import 'serialization.dart';

/// Webhook subcommand names (used for tab completion).
const webhookCommandNames = ['webhooks register', 'webhooks unregister', 'webhooks list'];

typedef WebhookHandler = Future<void> Function(BreezSdk sdk, List<String> args);

class _WebhookEntry {
  final String description;
  final WebhookHandler handler;
  const _WebhookEntry(this.description, this.handler);
}

Map<String, _WebhookEntry>? _registry;

Map<String, _WebhookEntry> _getRegistry() {
  return _registry ??= {
    'register': _WebhookEntry('Register a new webhook', _handleRegister),
    'unregister': _WebhookEntry('Unregister a webhook', _handleUnregister),
    'list': _WebhookEntry('List all registered webhooks', _handleList),
  };
}

/// Dispatch a webhook subcommand given the args after 'webhooks'.
Future<void> dispatchWebhookCommand(List<String> args, BreezSdk sdk) async {
  final registry = _getRegistry();

  if (args.isEmpty || args[0] == 'help' || args[0] == '--help') {
    print('\nWebhook subcommands:\n');
    for (final entry in registry.entries.toList()..sort((a, b) => a.key.compareTo(b.key))) {
      print('  webhooks ${entry.key.padRight(30)} ${entry.value.description}');
    }
    print('');
    return;
  }

  final subName = args[0];
  final subArgs = args.sublist(1);

  if (!registry.containsKey(subName)) {
    print("Unknown webhook subcommand: $subName. Use 'webhooks help' for available commands.");
    return;
  }

  await registry[subName]!.handler(sdk, subArgs);
}

WebhookEventType? _parseEventType(String s) {
  switch (s) {
    case 'lightning-receive':
      return WebhookEventType.lightningReceiveFinished();
    case 'lightning-send':
      return WebhookEventType.lightningSendFinished();
    case 'coop-exit':
      return WebhookEventType.coopExitFinished();
    case 'static-deposit':
      return WebhookEventType.staticDepositFinished();
    default:
      return null;
  }
}

// --- register ---

Future<void> _handleRegister(BreezSdk sdk, List<String> args) async {
  if (args.length < 3 || args.first == 'help' || args.first == '--help') {
    print('Usage: webhooks register <url> <secret> <event_type> [event_type...]');
    print('Event types: lightning-receive, lightning-send, coop-exit, static-deposit');
    return;
  }
  final url = args[0];
  final secret = args[1];
  final eventStrings = args.sublist(2);

  final eventTypes = <WebhookEventType>[];
  for (final e in eventStrings) {
    final eventType = _parseEventType(e);
    if (eventType == null) {
      print(
        'Unknown event type: $e. '
        'Valid values: lightning-receive, lightning-send, coop-exit, static-deposit',
      );
      return;
    }
    eventTypes.add(eventType);
  }

  final result = await sdk.registerWebhook(
    request: RegisterWebhookRequest(url: url, secret: secret, eventTypes: eventTypes),
  );
  printValue(result);
}

// --- unregister ---

Future<void> _handleUnregister(BreezSdk sdk, List<String> args) async {
  if (args.isEmpty || args.first == 'help' || args.first == '--help') {
    print('Usage: webhooks unregister <webhook_id>');
    return;
  }
  final webhookId = args[0];
  await sdk.unregisterWebhook(request: UnregisterWebhookRequest(webhookId: webhookId));
  print('Webhook unregistered successfully');
}

// --- list ---

Future<void> _handleList(BreezSdk sdk, List<String> args) async {
  final result = await sdk.listWebhooks();
  printValue(result);
}
